File: frontend/src/api/teams.ts
import { api } from './client';
import type { Team, TeamMessage } from '../stores/teams';
import type { BoardTask } from '../types/workspace';

// ---------------------------------------------------------------------------
// Backend response shapes
// ---------------------------------------------------------------------------

interface TeamListItem {
  name: string;
  state: string;
  member_count: number;
  created_at: string;
  parent_team: string | null;
  max_vms: number;
}

// team_status returns a rich report; we'll use the parts we need.
interface TeamStatusReport {
  name: string;
  state: string;
  members: TeamStatusMember[];
  max_vms: number;
  parent_team: string | null;
  created_at: string;
}

interface TeamStatusMember {
  name: string;
  role: string;
  skills: string[];
  workspace_id: string;
  ip: string;
  status: string;
}

// Backend BoardTask from vault-backed task board
interface TaskApiItem {
  id: string;
  title: string;
  description: string;
  status: string;
  owner: string | null;
  priority: string;
  depends_on: string[];
  created_at: string;
  completed_at: string | null;
}

// ---------------------------------------------------------------------------
// Conversion helpers
// ---------------------------------------------------------------------------

function mapTeamState(s: string): Team['state'] {
  // Backend uses Debug-format: "Creating", "Ready", etc.
  switch (s) {
    case 'Creating':
      return 'Creating';
    case 'Ready':
      return 'Ready';
    case 'Working':
      return 'Working';
    case 'Completing':
      return 'Completing';
    case 'Destroyed':
      return 'Destroyed';
    default:
      return 'Ready';
  }
}

function mapMemberStatus(s: string): 'idle' | 'busy' | 'offline' {
  switch (s.toLowerCase()) {
    case 'idle':
      return 'idle';
    case 'busy':
      return 'busy';
    case 'offline':
      return 'offline';
    default:
      return 'idle';
  }
}

function teamListItemToTeam(t: TeamListItem): Team {
  return {
    name: t.name,
    state: mapTeamState(t.state),
    members: [], // Members only available from team status endpoint
    max_vms: t.max_vms,
    parent_team: t.parent_team,
    created_at: t.created_at,
  };
}

function teamStatusToTeam(s: TeamStatusReport): Team {
  return {
    name: s.name,
    state: mapTeamState(s.state),
    members: s.members.map((m) => ({
      name: m.name,
      role: m.role,
      skills: m.skills,
      status: mapMemberStatus(m.status),
      workspace_id: m.workspace_id,
      ip: m.ip,
    })),
    max_vms: s.max_vms,
    parent_team: s.parent_team,
    created_at: s.created_at,
  };
}

function mapTaskStatus(s: string): BoardTask['status'] {
  switch (s) {
    case 'pending':
      return 'pending';
    case 'claimed':
      return 'claimed';
    case 'in_progress':
      return 'in_progress';
    case 'completed':
      return 'completed';
    case 'failed':
      return 'failed';
    default:
      return 'pending';
  }
}

function mapTaskPriority(s: string): BoardTask['priority'] {
  switch (s) {
    case 'low':
      return 'low';
    case 'high':
      return 'high';
    case 'critical':
      return 'critical';
    default:
      return 'medium';
  }
}

function taskApiToBoard(t: TaskApiItem, teamName: string): BoardTask {
  return {
    id: t.id,
    title: t.title,
    description: t.description,
    status: mapTaskStatus(t.status),
    owner: t.owner,
    priority: mapTaskPriority(t.priority),
    depends_on: t.depends_on,
    team_name: teamName,
    created_at: t.created_at,
    completed_at: t.completed_at,
  };
}

// ---------------------------------------------------------------------------
// API functions
// ---------------------------------------------------------------------------

export async function fetchTeams(): Promise<Team[]> {
  const items = await api.get<TeamListItem[]>('/teams');
  return items.map(teamListItemToTeam);
}

export async function fetchTeamStatus(name: string): Promise<Team> {
  const report = await api.get<TeamStatusReport>(
    `/teams/${encodeURIComponent(name)}`,
  );
  return teamStatusToTeam(report);
}

export async function fetchTeamTasks(teamName: string): Promise<BoardTask[]> {
  const items = await api.get<TaskApiItem[]>(
    `/teams/${encodeURIComponent(teamName)}/tasks`,
  );
  return items.map((t) => taskApiToBoard(t, teamName));
}

export async function fetchTeamMessages(
  teamName: string,
): Promise<TeamMessage[]> {
  return api.get<TeamMessage[]>(
    `/teams/${encodeURIComponent(teamName)}/messages`,
  );
}

export async function claimTask(
  teamName: string,
  taskId: string,
  agent: string | null,
): Promise<void> {
  await api.post(
    `/teams/${encodeURIComponent(teamName)}/tasks/${encodeURIComponent(taskId)}/claim`,
    { agent },
  );
}

export async function createTask(
  teamName: string,
  data: { title: string; description?: string; priority?: string; depends_on?: string[] },
): Promise<BoardTask> {
  const item = await api.post<TaskApiItem>(
    `/teams/${encodeURIComponent(teamName)}/tasks`,
    data,
  );
  return taskApiToBoard(item, teamName);
}

export async function startTask(
  teamName: string,
  taskId: string,
): Promise<void> {
  await api.post(
    `/teams/${encodeURIComponent(teamName)}/tasks/${encodeURIComponent(taskId)}/start`,
  );
}

export async function completeTask(
  teamName: string,
  taskId: string,
  result?: string,
): Promise<void> {
  await api.post(
    `/teams/${encodeURIComponent(teamName)}/tasks/${encodeURIComponent(taskId)}/complete`,
    result !== undefined ? { result } : {},
  );
}

export async function failTask(
  teamName: string,
  taskId: string,
  reason?: string,
): Promise<void> {
  await api.post(
    `/teams/${encodeURIComponent(teamName)}/tasks/${encodeURIComponent(taskId)}/fail`,
    reason !== undefined ? { reason } : {},
  );
}

export async function releaseTask(
  teamName: string,
  taskId: string,
): Promise<void> {
  await api.post(
    `/teams/${encodeURIComponent(teamName)}/tasks/${encodeURIComponent(taskId)}/release`,
  );
}

export async function sendTeamMessage(
  teamName: string,
  from: string,
  content: string,
  opts?: { to?: string; broadcast?: boolean },
): Promise<void> {
  await api.post(`/teams/${encodeURIComponent(teamName)}/messages`, {
    from,
    content,
    to: opts?.to,
    broadcast: opts?.broadcast ?? false,
  });
}

File: agentiso/src/dashboard/web/teams.rs
use std::sync::Arc;

use axum::{
    extract::{Path, State},
    response::Response,
    routing::{delete, get, post},
    Json, Router,
};
use serde::{Deserialize, Serialize};

use crate::team::RoleDef;
use crate::team::task_board::TaskPriority;

use super::{error_response, ok_response, DashboardState};

pub fn routes() -> Router<Arc<DashboardState>> {
    Router::new()
        .route("/teams", get(list_teams))
        .route("/teams", post(create_team))
        .route("/teams/{name}", get(get_team))
        .route("/teams/{name}", delete(destroy_team))
        .route("/teams/{name}/messages", get(get_messages))
        .route("/teams/{name}/messages", post(send_message))
        .route("/teams/{name}/tasks", get(list_tasks))
        .route("/teams/{name}/tasks", post(create_task))
        .route("/teams/{name}/tasks/{task_id}", get(get_task))
        .route("/teams/{name}/tasks/{task_id}/claim", post(claim_task))
        .route("/teams/{name}/tasks/{task_id}/start", post(start_task))
        .route("/teams/{name}/tasks/{task_id}/complete", post(complete_task))
        .route("/teams/{name}/tasks/{task_id}/fail", post(fail_task))
        .route("/teams/{name}/tasks/{task_id}/release", post(release_task))
}

// ---------------------------------------------------------------------------
// Request types
// ---------------------------------------------------------------------------

#[derive(Deserialize)]
struct CreateTeamRequest {
    name: String,
    roles: Vec<RoleDefRequest>,
    golden_workspace: Option<String>,
    base_snapshot: Option<String>,
    max_vms: Option<u32>,
    parent_team: Option<String>,
}

#[derive(Deserialize)]
struct RoleDefRequest {
    name: String,
    role: String,
    #[serde(default)]
    skills: Vec<String>,
    #[serde(default)]
    description: String,
}

#[derive(Deserialize)]
struct SendMessageRequest {
    from: String,
    to: Option<String>,
    content: String,
    #[serde(default)]
    broadcast: bool,
}

#[derive(Deserialize)]
struct CreateTaskRequest {
    title: String,
    description: Option<String>,
    priority: Option<String>,
    depends_on: Option<Vec<String>>,
}

#[derive(Deserialize)]
struct ClaimTaskRequest {
    owner: String,
}

#[derive(Deserialize)]
struct CompleteTaskRequest {
    result: Option<String>,
}

#[derive(Deserialize)]
struct FailTaskRequest {
    reason: Option<String>,
}

// ---------------------------------------------------------------------------
// Team response types
// ---------------------------------------------------------------------------

#[derive(Serialize)]
struct TeamInfo {
    name: String,
    state: String,
    member_count: usize,
    created_at: String,
    parent_team: Option<String>,
    max_vms: u32,
}

// ---------------------------------------------------------------------------
// Handlers
// ---------------------------------------------------------------------------

async fn list_teams(State(state): State<Arc<DashboardState>>) -> Response {
    let tm = match &state.team_manager {
        Some(tm) => tm,
        None => {
            return error_response(
                axum::http::StatusCode::NOT_FOUND,
                "TEAMS_DISABLED",
                "team management is not enabled",
            )
        }
    };

    match tm.list_teams().await {
        Ok(teams) => {
            let infos: Vec<TeamInfo> = teams
                .iter()
                .map(|t| TeamInfo {
                    name: t.name.clone(),
                    state: format!("{:?}", t.state),
                    member_count: t.member_workspace_ids.len(),
                    created_at: t.created_at.to_rfc3339(),
                    parent_team: t.parent_team.clone(),
                    max_vms: t.max_vms,
                })
                .collect();
            ok_response(infos)
        }
        Err(e) => error_response(
            axum::http::StatusCode::INTERNAL_SERVER_ERROR,
            "LIST_TEAMS_FAILED",
            e.to_string(),
        ),
    }
}

async fn create_team(
    State(state): State<Arc<DashboardState>>,
    Json(req): Json<CreateTeamRequest>,
) -> Response {
    let tm = match &state.team_manager {
        Some(tm) => tm,
        None => {
            return error_response(
                axum::http::StatusCode::NOT_FOUND,
                "TEAMS_DISABLED",
                "team management is not enabled",
            )
        }
    };

    let roles: Vec<RoleDef> = req
        .roles
        .into_iter()
        .map(|r| RoleDef {
            name: r.name,
            role: r.role,
            skills: r.skills,
            description: r.description,
        })
        .collect();

    // Resolve golden workspace ID if provided
    let golden_ws_id = if let Some(ref gw) = req.golden_workspace {
        match super::resolve_workspace_id(&state.workspace_manager, gw).await {
            Ok(id) => Some(id),
            Err((status, msg)) => return error_response(status, "NOT_FOUND", msg),
        }
    } else {
        None
    };

    let max_vms = req.max_vms.unwrap_or(20);

    match tm
        .create_team(
            &req.name,
            roles,
            golden_ws_id,
            req.base_snapshot.as_deref(),
            max_vms,
            req.parent_team.as_deref(),
        )
        .await
    {
        Ok(_team_state) => {
            use axum::response::IntoResponse;
            (
                axum::http::StatusCode::CREATED,
                axum::Json(super::ApiResponse::new(
                    serde_json::json!({"created": req.name}),
                )),
            )
                .into_response()
        }
        Err(e) => error_response(
            axum::http::StatusCode::BAD_REQUEST,
            "CREATE_TEAM_FAILED",
            e.to_string(),
        ),
    }
}

async fn get_team(
    State(state): State<Arc<DashboardState>>,
    Path(name): Path<String>,
) -> Response {
    let tm = match &state.team_manager {
        Some(tm) => tm,
        None => {
            return error_response(
                axum::http::StatusCode::NOT_FOUND,
                "TEAMS_DISABLED",
                "team management is not enabled",
            )
        }
    };

    match tm.team_status(&name).await {
        Ok(report) => ok_response(report),
        Err(e) => error_response(
            axum::http::StatusCode::NOT_FOUND,
            "TEAM_NOT_FOUND",
            e.to_string(),
        ),
    }
}

async fn destroy_team(
    State(state): State<Arc<DashboardState>>,
    Path(name): Path<String>,
) -> Response {
    let tm = match &state.team_manager {
        Some(tm) => tm,
        None => {
            return error_response(
                axum::http::StatusCode::NOT_FOUND,
                "TEAMS_DISABLED",
                "team management is not enabled",
            )
        }
    };

    match tm.destroy_team(&name).await {
        Ok(()) => ok_response(serde_json::json!({"destroyed": name})),
        Err(e) => error_response(
            axum::http::StatusCode::INTERNAL_SERVER_ERROR,
            "DESTROY_TEAM_FAILED",
            e.to_string(),
        ),
    }
}

async fn get_messages(
    State(state): State<Arc<DashboardState>>,
    Path(name): Path<String>,
) -> Response {
    // List all agents in this team and pull their messages
    let agents = state.message_relay.team_agents(&name).await;
    if agents.is_empty() {
        // Team might exist but no agents registered in relay
        return ok_response(serde_json::json!([]));
    }

    let mut all_messages = Vec::new();
    for agent_name in &agents {
        if let Ok(messages) = state.message_relay.receive(&name, agent_name, 50).await {
            all_messages.extend(messages);
        }
    }

    ok_response(all_messages)
}

async fn send_message(
    State(state): State<Arc<DashboardState>>,
    Path(name): Path<String>,
    Json(req): Json<SendMessageRequest>,
) -> Response {
    let to = if req.broadcast { "*" } else {
        match &req.to {
            Some(t) => t.as_str(),
            None => {
                return error_response(
                    axum::http::StatusCode::BAD_REQUEST,
                    "MISSING_RECIPIENT",
                    "set 'to' for direct message or 'broadcast: true' for team broadcast",
                );
            }
        }
    };

    match state
        .message_relay
        .send(&name, &req.from, to, &req.content, "text")
        .await
    {
        Ok(msg_id) => ok_response(serde_json::json!({"sent": true, "message_id": msg_id})),
        Err(e) => error_response(
            axum::http::StatusCode::BAD_REQUEST,
            "SEND_FAILED",
            e,
        ),
    }
}

// ---------------------------------------------------------------------------
// Task board handlers
// ---------------------------------------------------------------------------

fn parse_priority(s: &str) -> TaskPriority {
    match s {
        "low" => TaskPriority::Low,
        "high" => TaskPriority::High,
        "critical" => TaskPriority::Critical,
        _ => TaskPriority::Medium,
    }
}

async fn list_tasks(
    State(state): State<Arc<DashboardState>>,
    Path(name): Path<String>,
) -> Response {
    let vm = match &state.vault_manager {
        Some(vm) => vm,
        None => {
            return error_response(
                axum::http::StatusCode::NOT_FOUND,
                "VAULT_DISABLED",
                "vault is required for task board",
            )
        }
    };

    let board = crate::team::task_board::TaskBoard::new(vm.clone(), name);
    match board.list_tasks(None).await {
        Ok(tasks) => ok_response(tasks),
        Err(e) => error_response(
            axum::http::StatusCode::INTERNAL_SERVER_ERROR,
            "LIST_TASKS_FAILED",
            e.to_string(),
        ),
    }
}

async fn create_task(
    State(state): State<Arc<DashboardState>>,
    Path(name): Path<String>,
    Json(req): Json<CreateTaskRequest>,
) -> Response {
    let vm = match &state.vault_manager {
        Some(vm) => vm,
        None => {
            return error_response(
                axum::http::StatusCode::NOT_FOUND,
                "VAULT_DISABLED",
                "vault is required for task board",
            )
        }
    };

    let priority = parse_priority(req.priority.as_deref().unwrap_or("medium"));
    let depends_on = req.depends_on.unwrap_or_default();

    let board = crate::team::task_board::TaskBoard::new(vm.clone(), name);
    match board
        .create_task(
            &req.title,
            req.description.as_deref().unwrap_or(""),
            priority,
            depends_on,
        )
        .await
    {
        Ok(task) => {
            use axum::response::IntoResponse;
            (
                axum::http::StatusCode::CREATED,
                axum::Json(super::ApiResponse::new(task)),
            )
                .into_response()
        }
        Err(e) => error_response(
            axum::http::StatusCode::INTERNAL_SERVER_ERROR,
            "CREATE_TASK_FAILED",
            e.to_string(),
        ),
    }
}

async fn get_task(
    State(state): State<Arc<DashboardState>>,
    Path((name, task_id)): Path<(String, String)>,
) -> Response {
    let vm = match &state.vault_manager {
        Some(vm) => vm,
        None => {
            return error_response(
                axum::http::StatusCode::NOT_FOUND,
                "VAULT_DISABLED",
                "vault is required for task board",
            )
        }
    };

    let board = crate::team::task_board::TaskBoard::new(vm.clone(), name);
    match board.get_task(&task_id).await {
        Ok(task) => ok_response(task),
        Err(e) => error_response(
            axum::http::StatusCode::NOT_FOUND,
            "TASK_NOT_FOUND",
            e.to_string(),
        ),
    }
}

async fn claim_task(
    State(state): State<Arc<DashboardState>>,
    Path((name, task_id)): Path<(String, String)>,
    Json(req): Json<ClaimTaskRequest>,
) -> Response {
    let vm = match &state.vault_manager {
        Some(vm) => vm,
        None => {
            return error_response(
                axum::http::StatusCode::NOT_FOUND,
                "VAULT_DISABLED",
                "vault is required for task board",
            )
        }
    };

    let board = crate::team::task_board::TaskBoard::new(vm.clone(), name);
    match board.claim_task(&task_id, &req.owner).await {
        Ok(true) => ok_response(serde_json::json!({"claimed": task_id, "owner": req.owner})),
        Ok(false) => error_response(
            axum::http::StatusCode::CONFLICT,
            "ALREADY_CLAIMED",
            format!("task '{}' is already claimed", task_id),
        ),
        Err(e) => error_response(
            axum::http::StatusCode::CONFLICT,
            "CLAIM_FAILED",
            e.to_string(),
        ),
    }
}

async fn start_task(
    State(state): State<Arc<DashboardState>>,
    Path((name, task_id)): Path<(String, String)>,
) -> Response {
    let vm = match &state.vault_manager {
        Some(vm) => vm,
        None => {
            return error_response(
                axum::http::StatusCode::NOT_FOUND,
                "VAULT_DISABLED",
                "vault is required for task board",
            )
        }
    };

    let board = crate::team::task_board::TaskBoard::new(vm.clone(), name);
    match board.start_task(&task_id).await {
        Ok(()) => ok_response(serde_json::json!({"started": task_id})),
        Err(e) => error_response(
            axum::http::StatusCode::CONFLICT,
            "START_FAILED",
            e.to_string(),
        ),
    }
}

async fn complete_task(
    State(state): State<Arc<DashboardState>>,
    Path((name, task_id)): Path<(String, String)>,
    Json(req): Json<Option<CompleteTaskRequest>>,
) -> Response {
    let vm = match &state.vault_manager {
        Some(vm) => vm,
        None => {
            return error_response(
                axum::http::StatusCode::NOT_FOUND,
                "VAULT_DISABLED",
                "vault is required for task board",
            )
        }
    };

    let board = crate::team::task_board::TaskBoard::new(vm.clone(), name);
    let result_msg = req.and_then(|r| r.result);
    match board.complete_task(&task_id, result_msg.as_deref()).await {
        Ok(()) => ok_response(serde_json::json!({"completed": task_id})),
        Err(e) => error_response(
            axum::http::StatusCode::CONFLICT,
            "COMPLETE_FAILED",
            e.to_string(),
        ),
    }
}

async fn fail_task(
    State(state): State<Arc<DashboardState>>,
    Path((name, task_id)): Path<(String, String)>,
    Json(req): Json<Option<FailTaskRequest>>,
) -> Response {
    let vm = match &state.vault_manager {
        Some(vm) => vm,
        None => {
            return error_response(
                axum::http::StatusCode::NOT_FOUND,
                "VAULT_DISABLED",
                "vault is required for task board",
            )
        }
    };

    let board = crate::team::task_board::TaskBoard::new(vm.clone(), name);
    let reason = req.and_then(|r| r.reason).unwrap_or_default();
    match board.fail_task(&task_id, &reason).await {
        Ok(()) => ok_response(serde_json::json!({"failed": task_id})),
        Err(e) => error_response(
            axum::http::StatusCode::CONFLICT,
            "FAIL_FAILED",
            e.to_string(),
        ),
    }
}

async fn release_task(
    State(state): State<Arc<DashboardState>>,
    Path((name, task_id)): Path<(String, String)>,
) -> Response {
    let vm = match &state.vault_manager {
        Some(vm) => vm,
        None => {
            return error_response(
                axum::http::StatusCode::NOT_FOUND,
                "VAULT_DISABLED",
                "vault is required for task board",
            )
        }
    };

    let board = crate::team::task_board::TaskBoard::new(vm.clone(), name);
    match board.release_task(&task_id).await {
        Ok(()) => ok_response(serde_json::json!({"released": task_id})),
        Err(e) => error_response(
            axum::http::StatusCode::CONFLICT,
            "RELEASE_FAILED",
            e.to_string(),
        ),
    }
}
